### ISSUE TITLE
Don't recognize generators

### ISSUE THREAD

**Issue by DustinByfuglien** [rust-analyzer/rust-analyzer#4309]
```
#![feature(generators)]

fn main() {
    let x = || { yield 3 };
}
```

When using VSCode with this code I notice that:
1. It seems a ```yield``` with argument is reported by analyzer as error.
2. Also helping hint about ```x``` type displays that ```x``` have closure type ```|| -> i32``` instead of generator trait type.

**Comment by DustinByfuglien** [comment]
I can illustrate by screenshot.
1. Rust analyzer red underscored a error but no error is.
2. Hint shos that ```x``` have closure type ```|| -> i32``` but really it have generator type

![1](https://user-images.githubusercontent.com/18250092/81206866-63990480-8fd5-11ea-96d9-e59193de7b0d.png)

**Comment by ExoticMatter** [comment]
Interestingly, despite the bogus syntax error for no.1, rust-analyzer seems to typecheck yield arguments properly.

```rust
fn foo() {
    let bar = || {
        yield "baz";
        yield 0i32;     // E0308: expected `&str`, found `i32`
        yield;          // E0308: expected `&str`, found `()`
    };
}
```

![Demonstration](https://user-images.githubusercontent.com/7787686/81242237-fab88900-8fe2-11ea-91fb-2042d6b9bfad.png)

**Comment by flodiebold** [comment]
Any such error is coming from cargo check, not rust-analyzer's analysis.

**Comment by DustinByfuglien** [comment]
Cargo check passed without errors. Program compiled and executed without errors.

**Comment by flodiebold** [comment]
Sorry, I meant the E308 type checker errors.

**Comment by Veykril** [comment]
Parsing has been implemented in #7209, type inference is yet to be implemented.

**Comment by yuhr** [comment]
Type inference for the parameter should also be covered:

```rs
fn wrapper(mut generator: impl Generator<i32, Yield = i32> + Unpin) {
	while let Yielded(x) = Pin::new(&mut generator).resume(42) {
		println!("{}", x);
	}
}

fn main() {
	wrapper(|i| { // i: {unknown}
		yield i + 1;
		yield i + 2;
		yield i + 3;
	});
}
```
